dioptas/widgets/plot_widgets/utils.py

```diff
@@ -43,3 +45,133 @@ def auto_level(hist_x: np.ndarray, hist_y: np.ndarray) -> tuple[float, float]:
+    left_edges = bin_edges[:-1]
+    left_edges = left_edges[np.where(counts > 0)]
+    counts = counts[np.where(counts > 0)]
+
+    hist_y_cumsum = np.cumsum(counts)
+    hist_y_sum = np.sum(counts)
+
+    max_ind = np.where(hist_y_cumsum < (0.996 * hist_y_sum))
+    min_level = np.mean(left_edges[:2])
+
+    if len(max_ind[0]):
+        max_level = left_edges[max_ind[0][-1]]
+    else:
+        max_level = 0.5 * np.max(left_edges)
+
+    if len(left_edges[left_edges > 0]) > 0:
+        min_level = max(min_level, np.nanmin(left_edges[left_edges > 0]))
+    return min_level, max_level
+
+
+def _minmax_auto_level(data: np.ndarray) -> tuple[float, float]:
+    """Returns min/max of the data
+
+    :param data: Data from which to compute levels
+    :returns: (min, max)
+    """
+    return float(np.min(data)), float(np.max(data))
+
+
+def _mean3std_auto_level(data: np.ndarray) -> tuple[float, float]:
+    """Returns mean+/-3std clipped to min/max of the data
+
+    :param data: Data from which to compute levels
+    :returns: (lower limit, upper limit)
+    """
+    mean = np.mean(data, dtype=np.float64)
+    std = np.std(data, dtype=np.float64)
+    minimum, maximum = _minmax_auto_level(data)
+    return max(mean - 3 * std, minimum), min(mean + 3 * std, maximum)
+
+
+def _percentile_auto_level(
+    data: np.ndarray, percentile: float = 1.0
+) -> tuple[float, float]:
+    """Returns data range corresponding to [precentile, 100-percentile]
+
+    :param data: Data from which to compute levels
+    :returns: (lower limit, upper limit)
+    """
+    if not 0.0 <= percentile < 50.0:
+        raise ValueError("Percentiles must be in the range [0, 50[")
+    lower, upper = np.percentile(data, [percentile, 100.0 - percentile])
+    return float(lower), float(upper)
+
+
+def detector_gap_mask(data: np.ndarray) -> np.ndarray:
+    """Probe detector gap value and returns mask of pixels not belonging to gaps.
+
+    :param data: 2D image array for which to compute the mask
+    :returns: Mask with True for valid pixels and False for mask
+    """
+    if data.ndim != 2:
+        raise ValueError("2D array only are supported")
+
+    # Find columns with same values
+    equal_columns_mask = np.all(np.equal(data[0], data), axis=0)
+    if not np.any(equal_columns_mask):
+        return np.ones(data.shape, dtype=bool)
+
+    equal_values, counts = np.unique(data[0, equal_columns_mask], return_counts=True)
+    # At least 2 columns and not the whole image
+    mask = np.logical_and(counts > 1, counts < data.shape[1])
+    equal_values = equal_values[mask]
+    counts = counts[mask]
+    if len(equal_values) == 0:
+        return np.ones(data.shape, dtype=bool)
+
+    value = equal_values[np.argmax(counts)]
+    return data != value
+
+
+def auto_level(
+    data: Optional[np.ndarray],
+    hist_x: Optional[np.ndarray],
+    hist_y: Optional[np.ndarray],
+    mode: str = "default",
+    filter_gaps: bool = False,
+) -> Optional[tuple[float, float]]:
```

Comment:
This provides all the autoscale options.
It takes both image data and histogram as input. The histogram is only used by the current auto_levels mode, all the other are based on the image data.
In the end that could be simpler to only pass the data and compute the histogram here.